src/llama-graph.cpp

```diff
@@ -867,6 +874,14 @@ ggml_tensor * llm_graph_context::build_moe_ffn(
     }
 
     cur = ggml_reshape_3d(ctx0, cur, n_embd, 1, n_tokens);
+
+    if (weight_before_ffn) {
+        ggml_tensor * repeated = ggml_new_tensor_3d(ctx0, cur->type, n_embd, n_expert_used, n_tokens);
+        repeated = ggml_repeat(ctx0, cur, repeated); // [n_embd, n_expert_used, n_tokens]
+        cur = ggml_mul(ctx0, repeated, weights);
+        cb(cur, "ffn_moe_weighted", il);
+    }
```

Comment:
Not sure if this is the best way to repeat a dim. What I want to do here is to turn from:

`[n_embd, 1, n_tokens]`

to:

`[n_embd, n_expert_used, n_tokens]`

Reply:
In general, usage of `ggml_new_tensor_xd` should be avoided as much as possible because it can increase the required compute graph memory (i.e. allocate separate tensors for each layer, without reusing the memory). However, in this specific case, I can't think of an alternative way to do this.

Reply:
I tried ggml_mul directly but it failed the can_repeat check

Reply:
Um yeah we don't support broadcasting in both tensors at the same time. In that we should add a `ggml_repeat_4d` that just takes the desired dimension directly, instead of a tensor, to avoid creating a new tensor.

Reply:
An alternative approach for this particular case, we are currently doing:

```
cur = repeat(cur)
cur = mul(cur, weights)
gate = mm(cur, gate_w)
gate = silu(gate)
up = mm(cur, up_w)
```

But this also work, basically replace one `repeat` with 2 `mul`

```
gate = mul(mm(cur, gate_w), weights)
gate = silu(gate)
up = mul(mm(cur, up_w), weights)
```

But I think 2 times `ggml_mul` should be slower than 1 repeat and 1 mul